### ISSUE TITLE
Lowering Masking Operations to Linalg Depends on Mask Datatype

### ISSUE THREAD

**Issue by MaxS1996** [llvm/torch-mlir#4123]
I have encountered an issue with lowering masking operations to linalg, which seems to depend on the datatype of the mask. Below is a minimal example to reproduce the issue:

```python
import torch
import torch.nn as nn

class MaskingModule(nn.Module):
    def __init__(self, mask, dtype=torch.bool):
        super(MaskingModule, self).__init__()
        self.mask = mask.to(dtype)  # Convert mask to other types

    def forward(self, x):
        masked_output = torch.zeros_like(x)
        masked_output[self.mask] = x[self.mask]
        return masked_output

input_tensor = torch.tensor([[1, 2, 3], [4, 5, 6]], dtype=torch.float32)
mask = torch.tensor([[1, 0, 1], [0, 1, 0]], dtype=torch.float32)

dtype = torch.bool
model_torch = MaskingModule(mask, dtype=torch.bool)
output = model_torch(input_tensor)
```

**Steps to Reproduce:**

1. Use torch.int as the dtype for the mask: (`MaskingModule(mask, dtype=torch.int)`)

  *  Error during lowering to linalg: error: 'arith.cmpi' op requires all operands to have the same type
2. Use torch.long as the dtype for the mask: (`MaskingModule(mask, dtype=torch.long)`)

  * Lowering to linalg works as expected.

3. Use torch.bool as the dtype for the mask: (`MaskingModule(mask, dtype=torch.bool)`)

  * Lowering to linalg never terminates and continues to allocate more memory.

**Expected Behavior:**

Lowering should work consistently across different mask datatypes (torch itself supports long, int, byte or bool for the masking tensor) without causing errors or excessive memory allocation.

**Comment by MaxS1996** [comment]
Closer inspection seems to indicate that the problem seems to be related to IREE not torch-mlir
